R/runapp.R

```diff
@@ -135,11 +119,44 @@ runApp <- function(appDir=getwd(),
 
   require(shiny)
 
+  # ============================================================================
+  # Convert to Shiny app object
+  # ============================================================================
   appParts <- as.shiny.appobj(appDir)
 
+  # ============================================================================
+  # Initialize app state object
+  # ============================================================================
+  # This is so calls to getCurrentAppState() can be used to find (A) whether an
+  # app is running and (B), get options and data associated with the app.
+  initCurrentAppState(appParts)
+  on.exit(clearCurrentAppState(), add = TRUE)
+  # Any shinyOptions set after this point will apply to the current app only
+  # (and will not persist after the app stops).
+
+  # ============================================================================
+  # shinyOptions
+  # ============================================================================
+  # A unique identifier associated with this run of this application. It is
+  # shared across sessions.
+  shinyOptions(appToken = createUniqueId(8))
+
+  # Set up default cache for app.
+  if (is.null(getShinyOption("cache"))) {
+    shinyOptions(cache = MemoryCache$new())
+  }
+
+  # Extract appOptions (which is a list) and store them as shinyOptions, for
+  # this app. (This is the only place we have to store settings that are
+  # accessible both the UI and server portion of the app.)
+  applyCapturedAppOptions(appParts$appOptions)
```

Comment:
To me this would be clearer if this was just `if (!is.null(opts)) do.call(shinyOptions, opts)`, but I'm not sure if there's a good reason to use the helper elsewhere

Reply:
I think it's good to have the `applyCapturedAppOptions` function call just to make it clear that it's the counterpart to `captureAppOptions`. That function is also called from one other place, which is another reason to make it clear that it's the counterpart to `captureAppOptions`.

The purpose of this code is pretty confusing and it took me a while to wrap my head around it (even though I was the one who originally wrote it!), so anything to help is a good thing.